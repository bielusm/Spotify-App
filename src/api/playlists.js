import fetchToJson from "./util";

/**
 * Playlist
 * @description A container
 */
export default class Playlist {
  playlists = [];

  /**
   * [Gets information about user playlists and returns them  as a JSON object]
   *
   * @sync
   * @return {[promise]} [Contains playlist object JSON on success]
   *                     and API response on failure
   */
  async getPlaylists(bearerToken) {
    const json = await fetchToJson("https://api.spotify.com/v1/me/playlists", {
      headers: {
        Authorization: "Bearer " + bearerToken
      },
      method: "GET"
    });
    return json;
  }
  /**
   * Takes a ID and name of a playlist and saves/returns all of its tracks as JSON object
   *
   * @async
   * @param {int} ID   [the spotify ID of the playlist used to query the playlist]
   * @param {string} name [used to label the playlist]
   *
   * @return {Promise} Resolves with JSON on success, rejects with Error Object on failure
   */
  async addPlaylistByID(bearerToken, ID, name) {
    if (
      this.playlists.length > 0 &&
      this.playlists.find(element => element.id === ID)
    ) {
      return "already added";
    } else {
      let tracks = [];
      let param =
        "?fields=next,items(track(name,artists,external_urls,id,uri,album(name)))";
      let json = await fetchToJson(
        `https://api.spotify.com/v1/playlists/${ID}/tracks${param}`,
        {
          headers: {
            Authorization: "Bearer " + bearerToken
          },
          method: "GET"
        }
      );
      tracks.push(...json.items);
      const next = json["next"];
      if (next != null) {
        const response = await this.getNextTracks(
          bearerToken,
          ID,
          next,
          tracks
        );
        this.playlists.push({
          id: ID,
          tracks: tracks,
          name: name
        });
      } else {
        this.playlists.push({
          id: ID,
          tracks: tracks,
          name: name
        });
        return "success";
      }
    }
  }

  /**
   * Takes a URL from getNextTracks(ID, tracks) and fetches it,
   * this is used to deal with the fact that Spotify only returns 100 tracks at a time
   *
   * @async
   * @param {number} ID   ID of the Spotify playlist
   * @param {type} next   URL of the next playlist chunk to fetch
   * @param {type} tracks Array containing all tracks so far
   *
   * @return {type} Description
   */
  async getNextTracks(bearerToken, ID, next, tracks) {
    const json = await fetchToJson(next, {
      headers: {
        Authorization: "Bearer " + bearerToken
      },
      method: "GET"
    });
    tracks.push(...json.items);
    next = json["next"];
    if (next != null) return this.getNextTracks(bearerToken, ID, next, tracks);
    else return;
  }

  /**
   * Calls the currentTrackInPlaylist(index, track)
   * function with an index to each playlist, see that function below for more info
   * @param {track object} The track to be checked
   *
   *
   * @return {Array} An array of playlist IDs that contain the currents song
   */
  trackInPlaylists(track) {
    let ids = [];
    let index = 0;
    this.playlists.forEach(playlist => {
      if (this.trackInPlaylist(index, track)) ids.push(playlist);
      index++;
    });
    return ids;
  }

  /**
   * Checks if a playlist at the given index contains the current playing song
   *
   * @param {Number} index The index of the Spotify playlist
   * @param {Track Object} the track to be checked, needs a URI property
   * @return {Boolean} True if the track does contain the playlist, false otherwise
   */
  trackInPlaylist(index, track) {
    let equal = false;
    //TODO forEach is slow for this because it doesen't break the loop
    this.playlists[index].tracks.forEach(item => {
      if (item.track.uri === track.uri) equal = true;
    });
    return equal;
  }

  /**
   *  Adds a given track to a given playlist
   * @async
   * @param {String} playlist_id Spotify URI identifier for the playlist
   * @param {String} track_uri Spotify URI for the track to be added
   *
   * @return {Promise} Success of success and the error on failure
   */
  async addTrackToPlaylist(bearerToken, playlist_id, track) {
    const url = `https://api.spotify.com/v1/playlists/${playlist_id}/tracks?uris=${track.uri}`;
    const op = {
      headers: {
        Authorization: "Bearer " + bearerToken
      },
      method: "POST"
    };
    this.addTrackToCache(playlist_id, track);
    return fetchToJson(url, op);
  }

  /**
   * Takes a playlist URI and removes all instances of a given track URI from it
   * @async
   * @param {String} playlist_id
   * @param {String} track_uri
   *
   * @return {Promise} An empty promise
   */
  async removeTrackFromPlaylist(bearerToken, playlist_id, track) {
    const url = `https://api.spotify.com/v1/playlists/${playlist_id}/tracks`;
    const op = {
      headers: {
        Authorization: "Bearer " + bearerToken,
        "Content-Type": "application/json"
      },
      method: "DELETE",
      body: JSON.stringify({
        tracks: [
          {
            uri: track.uri
          }
        ]
      })
    };
    this.removeTrackFromCache(playlist_id, track);
    return await fetchToJson(url, op);
  }

  /**
   * Removes the playlist with the given playlist URI from the playlist array
   *
   * @param {Number} id  The Spotify URI for the playlist
   */
  removePlaylist(id) {
    const index = this.indexOfPlaylist(id);
    this.playlists.splice(index, 1);
  }

  /**
   * Finds the index of a playlist with the given playlist ID
   *
   * @param {Number} id
   */
  indexOfPlaylist(id) {
    for (let index = 0; index < this.playlists.length; index++) {
      if (this.playlists[index].id === id) return index;
    }
  }

  /**
   * Takes a track object and adds it to the array of all playlists
   *
   * @param {string} playlist_id the Spotify URI of the playlist
   * @param {Spotify track object} track
   */
  addTrackToCache(playlist_id, track) {
    const index = this.indexOfPlaylist(playlist_id);
    this.playlists[index].tracks.push({
      track
    });
  }

  /**
   * Takes a track object and removes it from the array of all playlists
   *
   * @param {string} playlist_id
   * @param {Spotify track object} track
   */
  removeTrackFromCache(playlist_id, track) {
    const index = this.indexOfPlaylist(playlist_id);
    this.playlists[index].tracks = this.playlists[index].tracks.filter(item => {
      return item.track.uri !== track.uri;
    });
  }
}
