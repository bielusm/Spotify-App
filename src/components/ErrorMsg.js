import React from "react";
import { connect } from "react-redux";

export const ErrorMsg = props => (
  <>
    {props.errMsg && (
      <h2 id="errMsg" className="my-4">
        {props.errMsg}
      </h2>
    )}
  </>
);

const mapStateToProps = state => ({ errMsg: state.error });

export default connect(mapStateToProps)(ErrorMsg);
