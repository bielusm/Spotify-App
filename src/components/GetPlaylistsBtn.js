import React from "react";

const GetPlaylistsBtn = props => (
  <button
    id="getPlaylistsBtn"
    className="button"
    onClick={props.getPlaylists}
    disabled={props.getPlaylistsDisabled}
  >
    Pick playlists
  </button>
);

export default GetPlaylistsBtn;
