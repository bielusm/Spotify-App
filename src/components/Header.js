import React from "react";

const Header = () => <h1>Spotify Manager</h1>;
export default Header;
