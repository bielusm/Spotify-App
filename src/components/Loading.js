import React from "react";

export const Loading = props => (
  <div>
    {props.loading && (
      <div className="loadingStatus">
        <div className="spinner-border" role="status">
          <span className="sr-only"></span>
        </div>
        <h2 className="mx-2 d-inline">Loading...</h2>
      </div>
    )}
  </div>
);

export default Loading;
