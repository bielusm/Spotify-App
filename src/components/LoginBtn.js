import React from "react";

export default class LoginBtn extends React.Component {
  login = () => {
    const scope =
      "playlist-modify-public%20user-read-playback-state%20playlist-read-private%20playlist-modify-private";
    const redirect_uri = window.location.origin;
    //redirect_uri = redirect_uri.slice(0, redirect_uri.length - 1);
    const url = `https://accounts.spotify.com/authorize?client_id=4252feb807d04ced962e15f346258957&response_type=token&redirect_uri=${redirect_uri}&scope=${scope}&show_dialog=false`;
    window.location.href = url;
  };

  render() {
    return (
      <div>
        {this.props.loginVisible && (
          <button id="login" className="btn btn-secondary" onClick={this.login}>
            Log in
          </button>
        )}
      </div>
    );
  }
}
