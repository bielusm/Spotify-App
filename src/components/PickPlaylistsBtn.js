import React from "react";
import Button from "@material-ui/core/Button";

export const PickPlaylistsBtn = props => (
  <Button
    variant="contained"
    color="primary"
    id="pickPlaylistsBtn"
    onClick={props.getPlaylists}
    disabled={props.getPlaylistsDisabled}
  >
    Pick playlists
  </Button>
);

export default PickPlaylistsBtn;
