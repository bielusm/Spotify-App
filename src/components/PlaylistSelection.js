import React from "react";

export const PlaylistSelection = props => (
  <>
    {props.showPlaylists && (
      <div>
        <h2 className="playlist-selection__header">Your playlists:</h2>
        <div
          id="playlist-selection"
          className="playlist-selection__list"
          onClick={props.selectPlaylist}
        >
          {props.playlists.map(playlist => {
            const plName = playlist.name.replace(/ /g, "-");
            const selected = playlist.selected;
            let className = "playlist-selection__item";
            if (selected) className += " active";
            return (
              <div key={playlist.name} id={playlist.id} className={className}>
                {playlist.name}
              </div>
            );
          })}
        </div>
      </div>
    )}
  </>
);

export default PlaylistSelection;
