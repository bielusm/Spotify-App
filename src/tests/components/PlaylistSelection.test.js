import React from "react";
import { shallow } from "enzyme";
import { PlaylistSelection } from "../../components/PlaylistSelection";
import { playlists as playlistsFixture } from "../fixtures/playlists";

let showPlaylists, selectPlaylist, wrapper, playlists;
beforeEach(() => {
  showPlaylists = true;
  selectPlaylist = jest.fn();
  playlists = playlistsFixture;
  wrapper = shallow(
    <PlaylistSelection
      showPlaylists={showPlaylists}
      playlists={playlists}
      selectPlaylist={selectPlaylist}
    />
  );
});

test("Snapshot with showPlaylists false", () => {
  wrapper.setProps({ showPlaylists: false });
  expect(wrapper).toMatchSnapshot();
});

test("Snapshot with showPlaylists true", () => {
  wrapper.setProps({ showPlaylists: true });
  expect(wrapper).toMatchSnapshot();
});
