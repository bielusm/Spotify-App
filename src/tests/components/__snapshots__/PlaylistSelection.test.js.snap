// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot with showPlaylists false 1`] = `<Fragment />`;

exports[`Snapshot with showPlaylists true 1`] = `
<Fragment>
  <div>
    <h2
      className="playlist-selection__header"
    >
      Your playlists:
    </h2>
    <div
      className="playlist-selection__list"
      id="playlist-selection"
      onClick={[MockFunction]}
    >
      <div
        className="playlist-selection__item"
        id="3F94pf3iVnCOSe0lvHUuIl"
        key="New Playlist 2"
      >
        New Playlist 2
      </div>
      <div
        className="playlist-selection__item"
        id="3iKeac3IvswDw05mv1dc3z"
        key="api testing"
      >
        api testing
      </div>
      <div
        className="playlist-selection__item"
        id="43x3DlIWcCJzEBNEBEDHHh"
        key="piano"
      >
        piano
      </div>
      <div
        className="playlist-selection__item"
        id="4TL6CsXn7AFFrgMWMxq1XL"
        key="Big rap playlist"
      >
        Big rap playlist
      </div>
      <div
        className="playlist-selection__item"
        id="6R4CXv7L5mETXDD3z9360j"
        key="everything"
      >
        everything
      </div>
      <div
        className="playlist-selection__item"
        id="75fBy2AgLDGKoOyCfYwGSG"
        key="calm"
      >
        calm
      </div>
      <div
        className="playlist-selection__item"
        id="2PnJuP3B7Lm2DPs8nEqpiq"
        key="bad programming playlist"
      >
        bad programming playlist
      </div>
      <div
        className="playlist-selection__item"
        id="2lcW5jiLBwniHL8CyKg1H8"
        key=":O"
      >
        :O
      </div>
      <div
        className="playlist-selection__item"
        id="23cUfHhE4yOIHGEpCus559"
        key="top"
      >
        top
      </div>
      <div
        className="playlist-selection__item"
        id="37i9dQZEVXcIBiEFq8ePzj"
        key="Discover Weekly"
      >
        Discover Weekly
      </div>
      <div
        className="playlist-selection__item"
        id="37i9dQZEVXblmn3RrluvDS"
        key="Release Radar"
      >
        Release Radar
      </div>
      <div
        className="playlist-selection__item"
        id="6SFU42MrxusLKBM2dywtmY"
        key="Go through"
      >
        Go through
      </div>
      <div
        className="playlist-selection__item"
        id="6uuht6sd1Rm5lFvAoRgzAj"
        key="mishmash"
      >
        mishmash
      </div>
      <div
        className="playlist-selection__item"
        id="1A7bJ7EKtT9XqpN0yV9qCf"
        key="drums"
      >
        drums
      </div>
    </div>
  </div>
</Fragment>
`;
